components/script/dom/texttracklist.rs

```diff
@@ -0,0 +1,101 @@
+use crate::dom::texttrack::TextTrack;
+use crate::dom::window::Window;
+use dom_struct::dom_struct;
+
+#[dom_struct]
+pub struct TextTrackList {
+    eventtarget: EventTarget,
+    dom_tracks: DomRefCell<Vec<Dom<TextTrack>>>,
+}
+
+impl TextTrackList {
+    pub fn new_inherited(tracks: &[&TextTrack]) -> TextTrackList {
+        TextTrackList {
+            eventtarget: EventTarget::new_inherited(),
+            dom_tracks: DomRefCell::new(tracks.iter().map(|g| Dom::from_ref(&**g)).collect()),
+        }
+    }
+
+    pub fn new(window: &Window, tracks: &[&TextTrack]) -> DomRoot<TextTrackList> {
+        reflect_dom_object(
+            Box::new(TextTrackList::new_inherited(tracks)),
+            window,
+            TextTrackListBinding::Wrap,
+        )
+    }
+
+    pub fn item(&self, idx: usize) -> Option<DomRoot<TextTrack>> {
+        self.dom_tracks
+            .borrow()
+            .get(idx as usize)
+            .map(|t| DomRoot::from_ref(&**t))
+    }
+
+    pub fn find(&self, track: &TextTrack) -> Option<usize> {
+        self.dom_tracks
+            .borrow()
+            .iter()
+            .enumerate()
+            .filter(|(_, t)| **t == track)
+            .next()
+            .map(|(i, _)| i)
+    }
+
+    pub fn add(&self, track: &TextTrack) {
+        // Only add a track if it does not exist in the list
+        if self.find(track).is_none() {
+            self.dom_tracks.borrow_mut().push(Dom::from_ref(track))
+        };
+        self.upcast::<EventTarget>()
+            .fire_event(atom!("addtrack"));
+    }
+
+    pub fn remove(&self, idx: usize) {
+        self.dom_tracks.borrow_mut().remove(idx);
+        self.upcast::<EventTarget>()
+            .fire_event(atom!("removetrack"));
+    }
+}
+
+impl TextTrackListMethods for TextTrackList {
+    // https://html.spec.whatwg.org/multipage/#dom-texttracklist-length
+    fn Length(&self) -> u32 {
+        self.dom_tracks.borrow().len() as u32
+    }
+
+    // https://html.spec.whatwg.org/multipage/#dom-texttracklist-item
+    fn IndexedGetter(&self, idx: u32) -> Option<DomRoot<TextTrack>> {
+        self.item(idx as usize)
+    }
+
+    // https://html.spec.whatwg.org/multipage/#dom-texttracklist-gettrackbyid
+    fn GetTrackById(&self, id: DOMString) -> Option<DomRoot<TextTrack>> {
+        let id_str = String::from(id.clone());
+        self.dom_tracks
+            .borrow()
+            .iter()
+            .filter(|track| track.id() == &id_str)
+            .next()
+            .map(|t| DomRoot::from_ref(&**t))
+    }
+
+    // https://html.spec.whatwg.org/multipage/#handler-texttracklist-onchange
+    event_handler!(onchange, GetOnchange, SetOnchange);
+
+    // https://html.spec.whatwg.org/multipage/#handler-texttracklist-onaddtrack
+    event_handler!(onchange, GetOnaddtrack, SetOnaddtrack);
+
+    // https://html.spec.whatwg.org/multipage/#handler-texttracklist-onremovetrack
+    event_handler!(onchange, GetOnremovetrack, SetOnremovetrack);
```

Comment:
These are using onchange, which probably explains why a few test failures remain in the event handling tests.

Reply:
Good catch. [TrackEvent](https://html.spec.whatwg.org/multipage/media.html#trackevent) hasn't been implemented yet, so we'll likely have to implement that in order to get these to work correctly right? I haven't looked into the internals of events yet, so I'm very likely wrong :smile: